src/persist/src/indexed/columnar/parquet.rs

```diff
@@ -0,0 +1,171 @@
+
+use arrow2::io::parquet::read::RecordReader;
+use arrow2::io::parquet::write::RowGroupIterator;
+use arrow2::record_batch::RecordBatch;
+use differential_dataflow::trace::Description;
+use parquet2::compression::Compression;
+use parquet2::encoding::Encoding;
+use parquet2::metadata::KeyValue;
+use parquet2::write::{write_file, Version, WriteOptions};
+use timely::progress::{Antichain, Timestamp};
+
+use crate::error::Error;
+use crate::gen::persist::ProtoBatchFormat;
+use crate::indexed::columnar::arrow::SCHEMA_ARROW_KVTD;
+use crate::indexed::columnar::ColumnarRecords;
+use crate::indexed::encoding::{
+    decode_trace_inline, decode_unsealed_inline, encode_trace_inline_meta,
+    encode_unsealed_inline_meta, BlobTraceBatch, BlobUnsealedBatch,
+};
+use crate::storage::SeqNo;
+
+const INLINE_METADATA_KEY: &'static str = "MZ:meta";
+
+/// Encodes an BlobUnsealedBatch into the Parquet format.
+pub fn encode_unsealed_parquet<W: Write>(
+    w: &mut W,
+    batch: &BlobUnsealedBatch,
+) -> Result<(), Error> {
+    encode_parquet_kvtd(
+        w,
+        encode_unsealed_inline_meta(batch, ProtoBatchFormat::ParquetKVTD),
+        &batch.updates,
+    )
+}
+
+/// Encodes an BlobTraceBatch into the Parquet format.
+pub fn encode_trace_parquet<W: Write>(w: &mut W, batch: &BlobTraceBatch) -> Result<(), Error> {
+    let updates = batch
+        .updates
+        .iter()
+        .map(|((k, v), t, d)| ((k.to_vec(), v.to_vec()), *t, *d))
+        .collect();
+    encode_parquet_kvtd(
+        w,
+        encode_trace_inline_meta(batch, ProtoBatchFormat::ParquetKVTD),
+        &[updates],
+    )
+}
+
+/// Decodes a BlobUnsealedBatch from the Parquet format.
+pub fn decode_unsealed_parquet<R: Read + Seek>(r: &mut R) -> Result<BlobUnsealedBatch, Error> {
+    let reader = RecordReader::try_new(r, None, None, None, None)?;
+    let metadata = reader
+        .metadata()
+        .key_value_metadata()
+        .as_ref()
+        .and_then(|x| x.iter().find(|x| x.key == INLINE_METADATA_KEY));
+    let (format, meta) = decode_unsealed_inline(metadata.and_then(|x| x.value.as_ref()))?;
+
+    let updates = match format {
+        ProtoBatchFormat::Unknown => return Err("unknown format".into()),
+        ProtoBatchFormat::ArrowKVTD => {
+            return Err("ArrowKVTD format not supported in parquet".into())
+        }
+        ProtoBatchFormat::ParquetKVTD => {
+            let mut updates = Vec::new();
+            for batch in reader {
+                updates.push(ColumnarRecords::try_from(batch?)?);
+            }
+            updates
+        }
+    };
+
+    let ret = BlobUnsealedBatch {
+        desc: SeqNo(meta.seqno_lower)..SeqNo(meta.seqno_upper),
+        updates,
+    };
+    ret.validate()?;
+    Ok(ret)
+}
+
+/// Decodes a BlobTraceBatch from the Parquet format.
+pub fn decode_trace_parquet<R: Read + Seek>(r: &mut R) -> Result<BlobTraceBatch, Error> {
+    let reader = RecordReader::try_new(r, None, None, None, None)?;
```

Comment:
note to self: probably want to be more explicit about checking that the schema matches the format like we do in the arrow equivalents